File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) 
    {
        int c[3] ={},ans=0,i=0,j=0;
        int n = s.length();
        for(;j<n;j++) 
        {
            c[s[j] - 'a']++;
            while (c[0] && c[1] && c[2])
            c[s[i++] - 'a']--;
            ans+=i;
        }
        return ans;
    }
};

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.js
//translated using AI
/**
 * @param {string} s
 * @return {number}
 */
var numberOfSubstrings = function(s) {
    let lastA = -1, lastB = -1, lastC = -1;
    let total = 0;

    for (let i = 0; i < s.length; i++) {
        if (s[i] === 'a')
            lastA = i;
        else if (s[i] === 'b')
            lastB = i;
        else
            lastC = i;

        if (lastA !== -1 && lastB !== -1 && lastC !== -1) {
            total += Math.min(lastA, lastB, lastC) + 1;
        }
    }

    return total;
};